activesupport/lib/active_support/hash_with_indifferent_access.rb

```diff
@@ -78,7 +78,7 @@ def initialize(constructor = {})
     def default(*args)
       arg_key = args.first
 
-      if include?(key = convert_key(arg_key))
+      if include?(key = convert_key(arg_key)) && args.size > 0
         self[key]
```

Comment:
Hmm... sounds like we need to define our own `dig`: the one from the superclass [uses rb_hash_aref directly](https://github.com/ruby/ruby/blob/2430822b96a399ce5983b26637f597019690eb3e/hash.c#L2878), so doesn't get the `convert_key` treatment. 

It currently then "works" because after the aref misses, it falls back to the default value, which we lie about here to turn it into a second-chance lookup-by-key -- which is how all lookups worked prior to #18909.

So yeah, I think #23414 was just the wrong fix for the reported issue: #18909 should have added `dig` at the same time it added `[]` (but presumably missed it because `dig` was super-new at the time, IIRC).